argh/src/lib.rs

```diff
@@ -166,6 +166,95 @@
 //!     fooey: bool,
 //! }
 //! ```
+//!
+//! You can also discover subcommands dynamically at runtime. To do this, declare subcommands as
+//! usual and add a variant to the enum with the `dynamic` attribute. Instead of deriving
+//! `FromArgs`, the value inside the dynamic variant should implement `DynamicSubCommand`.
+//!
+//! ```rust
+//! # use argh::CommandInfo;
+//! # use argh::DynamicSubCommand;
+//! # use argh::EarlyExit;
+//! # use argh::FromArgs;
+//! # use once_cell::sync::OnceCell;
+//!
+//! #[derive(FromArgs, PartialEq, Debug)]
+//! /// Top-level command.
+//! struct TopLevel {
+//!     #[argh(subcommand)]
+//!     nested: MySubCommandEnum,
+//! }
+//!
+//! #[derive(FromArgs, PartialEq, Debug)]
+//! #[argh(subcommand)]
+//! enum MySubCommandEnum {
+//!     Normal(NormalSubCommand),
+//!     #[argh(dynamic)]
+//!     Dynamic(Dynamic),
+//! }
+//!
+//! #[derive(FromArgs, PartialEq, Debug)]
+//! /// Normal subcommand.
+//! #[argh(subcommand, name = "normal")]
+//! struct NormalSubCommand {
+//!     #[argh(option)]
+//!     /// how many x
+//!     x: usize,
+//! }
+//!
+//! /// Dynamic subcommand.
+//! #[derive(PartialEq, Debug)]
+//! struct Dynamic {
+//!     name: String
+//! }
+//!
+//! impl DynamicSubCommand for Dynamic {
+//!     fn commands() -> &'static [&'static CommandInfo] {
+//!         static RET: OnceCell<Vec<&'static CommandInfo>> = OnceCell::new();
+//!         RET.get_or_init(|| {
+//!             let mut commands = Vec::new();
+//!
+//!             // Do whatever dynamic process you want to populate your commands list here.
+//!             commands.push(&*Box::leak(Box::new(CommandInfo {
+//!                 name: "dynamic_command",
```

Comment:
I think we either need to simplify this example a little bit, or explain it a little more. It looks like even your test cases aren't using `Box::leak()`. (Perhaps they should, so someone can, if they want to, copy you're test code as a starting point for their own implementation--assuming most use cases really need to leak a dynamic subcommand list in order to support the static lifetime references.) I'm assuming there isn't a _simpler_ way, and my concern is that most developers rarely if ever need to use Box::leak(). I think I understand why, but I think it would be helpful, for less advanced Rust programmers, to explain what you're doing.

The triple static lifetime referencing is confusing in its own right. Again, probably necessary to show a useful use case, but it's hard to follow.

Anything you can do to make this more clear would be great.

Reply:
Made the comment a bit longer to explain some of these things.